File: lib/presentation/views/circles/sos_prayer_request_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../../data/datasources/remote/auth_service.dart';
import '../../../domain/entities/circle.dart';
import '../../../domain/repositories/circle_repository.dart';
import '../../theme/app_theme.dart';

class SOSPrayerRequestView extends StatefulWidget {
  final String circleId;
  final List<CircleMember> members;

  const SOSPrayerRequestView({super.key, required this.circleId, required this.members});

  @override
  State<SOSPrayerRequestView> createState() => _SOSPrayerRequestViewState();
}

class _SOSPrayerRequestViewState extends State<SOSPrayerRequestView> {
  final _messageController = TextEditingController();
  final Set<String> _selectedIds = {};
  bool _isSending = false;
  String? _error;
  bool _sentSuccessfully = false;
  int _recipientCount = 0;

  static const _maxRecipients = 20;

  List<CircleMember> get _otherMembers {
    final myId = AuthService.shared.userId;
    return widget.members.where((m) => m.userId != myId).toList();
  }

  @override
  void dispose() {
    _messageController.dispose();
    super.dispose();
  }

  Future<void> _sendSOS() async {
    setState(() { _isSending = true; _error = null; });
    final msg = _messageController.text.trim();
    final finalMsg = msg.isEmpty ? 'Please pray for me' : msg;
    try {
      final count = _selectedIds.length;
      await context.read<CircleRepository>().sendSOS(
          widget.circleId, finalMsg, _selectedIds.toList());
      if (!mounted) return;
      setState(() { _recipientCount = count; _sentSuccessfully = true; });
    } catch (e) {
      if (mounted) setState(() { _error = e.toString(); _isSending = false; });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: MyWalkColor.charcoal,
      appBar: AppBar(
        backgroundColor: MyWalkColor.charcoal,
        title: Text(_sentSuccessfully ? '' : 'SOS Prayer',
            style: const TextStyle(color: MyWalkColor.warmWhite, fontSize: 17, fontWeight: FontWeight.w600)),
        leading: TextButton(
          onPressed: () => Navigator.pop(context),
          child: Text(_sentSuccessfully ? 'Done' : 'Cancel',
              style: TextStyle(color: Colors.white.withValues(alpha: 0.5))),
        ),
        leadingWidth: 80,
      ),
      body: SafeArea(child: _sentSuccessfully ? _successView() : _composeView()),
    );
  }

  Widget _successView() {
    return Column(mainAxisAlignment: MainAxisAlignment.center, children: [
      Container(
        width: 88, height: 88,
        decoration: BoxDecoration(shape: BoxShape.circle, color: MyWalkColor.sage.withValues(alpha: 0.12)),
        child: const Icon(Icons.back_hand_rounded, size: 36, color: MyWalkColor.sage),
      ),
      const SizedBox(height: 24),
      const Text('Prayer Request Sent',
          style: TextStyle(fontSize: 20, fontWeight: FontWeight.w700, color: MyWalkColor.warmWhite)),
      const SizedBox(height: 8),
      Text(
        '$_recipientCount ${_recipientCount == 1 ? 'person has' : 'people have'} been asked to pray for you.',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 15, color: Colors.white.withValues(alpha: 0.6)),
      ),
      const SizedBox(height: 24),
      const Padding(
        padding: EdgeInsets.symmetric(horizontal: 32),
        child: Text(
          '"Bear one another\'s burdens, and so fulfill the law of Christ."',
          textAlign: TextAlign.center,
          style: TextStyle(fontSize: 14, fontStyle: FontStyle.italic, color: MyWalkColor.softGold),
        ),
      ),
      const SizedBox(height: 8),
      Text('Galatians 6:2',
          style: TextStyle(fontSize: 12, color: MyWalkColor.golden.withValues(alpha: 0.5))),
    ]);
  }

  Widget _composeView() {
    final others = _otherMembers;
    final allSelected = others.isNotEmpty &&
        _selectedIds.length == others.take(_maxRecipients).length;

    return SingleChildScrollView(
      padding: const EdgeInsets.fromLTRB(20, 16, 20, 40),
      child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
        Center(
          child: Column(children: [
            Container(
              width: 64, height: 64,
              decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: MyWalkColor.warmCoral.withValues(alpha: 0.12)),
              child: const Icon(Icons.campaign_rounded, size: 28, color: MyWalkColor.warmCoral),
            ),
            const SizedBox(height: 12),
            const Text('Request Urgent Prayer',
                style: TextStyle(fontSize: 18, fontWeight: FontWeight.w700, color: MyWalkColor.warmWhite)),
            const SizedBox(height: 4),
            Text('Select up to $_maxRecipients people who will be notified to pray for you.',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.5))),
          ]),
        ),
        const SizedBox(height: 24),
        Row(children: [
          const Text('Message',
              style: TextStyle(fontSize: 12, fontWeight: FontWeight.w600, color: MyWalkColor.softGold)),
          const Spacer(),
          Text('${_messageController.text.length}/500',
              style: TextStyle(fontSize: 11, color: Colors.white.withValues(alpha: 0.3))),
        ]),
        const SizedBox(height: 6),
        TextField(
          controller: _messageController,
          onChanged: (_) => setState(() {}),
          maxLines: 4,
          maxLength: 500,
          buildCounter: (_, {required currentLength, required isFocused, maxLength}) => null,
          style: const TextStyle(color: MyWalkColor.warmWhite),
          decoration: InputDecoration(
            hintText: 'Please pray for me...',
            hintStyle: TextStyle(color: Colors.white.withValues(alpha: 0.3)),
            filled: true, fillColor: MyWalkColor.cardBackground,
            border: OutlineInputBorder(borderRadius: BorderRadius.circular(10),
                borderSide: BorderSide(color: MyWalkColor.cardBorder, width: 0.5)),
            enabledBorder: OutlineInputBorder(borderRadius: BorderRadius.circular(10),
                borderSide: BorderSide(color: MyWalkColor.cardBorder, width: 0.5)),
            focusedBorder: OutlineInputBorder(borderRadius: BorderRadius.circular(10),
                borderSide: BorderSide(color: MyWalkColor.golden.withValues(alpha: 0.5), width: 1)),
            contentPadding: const EdgeInsets.all(12),
          ),
        ),
        const SizedBox(height: 20),
        Row(children: [
          const Text('Recipients',
              style: TextStyle(fontSize: 12, fontWeight: FontWeight.w600, color: MyWalkColor.softGold)),
          const Spacer(),
          Text('${_selectedIds.length}/$_maxRecipients',
              style: TextStyle(
                  fontSize: 11, fontWeight: FontWeight.w500,
                  color: _selectedIds.length >= _maxRecipients
                      ? MyWalkColor.warmCoral
                      : Colors.white.withValues(alpha: 0.4))),
        ]),
        const SizedBox(height: 8),
        if (others.isEmpty)
          Container(
            padding: const EdgeInsets.all(12),
            decoration: BoxDecoration(color: MyWalkColor.cardBackground, borderRadius: BorderRadius.circular(10)),
            child: Row(children: [
              Icon(Icons.person_off_rounded, size: 14, color: Colors.white.withValues(alpha: 0.4)),
              const SizedBox(width: 8),
              Text('No other members in this circle yet.',
                  style: TextStyle(fontSize: 12, color: Colors.white.withValues(alpha: 0.4))),
            ]),
          )
        else ...[
          TextButton(
            onPressed: () => setState(() {
              if (allSelected) {
                _selectedIds.clear();
              } else {
                _selectedIds.addAll(others.take(_maxRecipients).map((m) => m.userId));
              }
            }),
            style: TextButton.styleFrom(padding: EdgeInsets.zero),
            child: Row(children: [
              Icon(allSelected ? Icons.check_circle_rounded : Icons.circle_outlined,
                  size: 18,
                  color: allSelected ? MyWalkColor.golden : Colors.white.withValues(alpha: 0.4)),
              const SizedBox(width: 8),
              Text(
                allSelected
                    ? 'Deselect All'
                    : 'Select All (${others.take(_maxRecipients).length})',
                style: const TextStyle(fontSize: 13, color: MyWalkColor.warmWhite),
              ),
            ]),
          ),
          ...others.map((m) => Padding(
                padding: const EdgeInsets.only(bottom: 8),
                child: _recipientRow(m),
              )),
        ],
        if (_error != null) ...[
          const SizedBox(height: 12),
          Row(children: [
            const Icon(Icons.warning_amber, size: 14, color: MyWalkColor.warmCoral),
            const SizedBox(width: 6),
            Expanded(child: Text(_error!, style: const TextStyle(fontSize: 12, color: MyWalkColor.warmCoral))),
          ]),
        ],
        const SizedBox(height: 24),
        Opacity(
          opacity: _selectedIds.isEmpty ? 0.5 : 1.0,
          child: SizedBox(
            width: double.infinity,
            child: ElevatedButton.icon(
              onPressed: (_selectedIds.isEmpty || _isSending) ? null : _sendSOS,
              icon: _isSending
                  ? const SizedBox(width: 18, height: 18,
                      child: CircularProgressIndicator(strokeWidth: 2, color: MyWalkColor.charcoal))
                  : const Icon(Icons.bolt_rounded, size: 18),
              label: const Text('Send SOS Prayer Request',
                  style: TextStyle(fontWeight: FontWeight.w600, fontSize: 16)),
              style: ElevatedButton.styleFrom(
                backgroundColor: MyWalkColor.warmCoral,
                foregroundColor: MyWalkColor.charcoal,
                padding: const EdgeInsets.symmetric(vertical: 16),
                shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(14)),
              ),
            ),
          ),
        ),
      ]),
    );
  }

  Widget _recipientRow(CircleMember m) {
    final isSelected = _selectedIds.contains(m.userId);
    final isDisabled = !isSelected && _selectedIds.length >= _maxRecipients;
    return Opacity(
      opacity: isDisabled ? 0.4 : 1.0,
      child: GestureDetector(
        onTap: isDisabled ? null : () => setState(() {
          if (isSelected) {
            _selectedIds.remove(m.userId);
          } else {
            _selectedIds.add(m.userId);
          }
        }),
        child: Container(
          padding: const EdgeInsets.all(10),
          decoration: BoxDecoration(
            color: isSelected ? MyWalkColor.golden.withValues(alpha: 0.04) : MyWalkColor.cardBackground,
            borderRadius: BorderRadius.circular(10),
            border: Border.all(
              color: isSelected ? MyWalkColor.golden.withValues(alpha: 0.2) : MyWalkColor.cardBorder,
              width: 0.5,
            ),
          ),
          child: Row(children: [
            Container(
              width: 36, height: 36,
              decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  color: isSelected ? MyWalkColor.golden.withValues(alpha: 0.15) : MyWalkColor.cardBackground),
              child: Icon(Icons.person_rounded, size: 14,
                  color: isSelected ? MyWalkColor.golden : Colors.white.withValues(alpha: 0.5)),
            ),
            const SizedBox(width: 12),
            Expanded(child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
              const Text('Member',
                  style: TextStyle(fontSize: 14, fontWeight: FontWeight.w500, color: MyWalkColor.warmWhite)),
              if (m.role == 'admin')
                const Text('Admin', style: TextStyle(fontSize: 11, color: MyWalkColor.golden)),
            ])),
            Icon(isSelected ? Icons.check_circle_rounded : Icons.circle_outlined,
                size: 22,
                color: isSelected ? MyWalkColor.golden : Colors.white.withValues(alpha: 0.15)),
          ]),
        ),
      ),
    );
  }
}

File: swift_reference/Views/SOSPrayerRequestView.swift
import SwiftUI

struct SOSPrayerRequestView: View {
    let circleId: String
    let members: [CircleMemberInfo]

    @Environment(\.dismiss) private var dismiss
    @State private var message: String = ""
    @State private var selectedRecipientIds: Set<String> = []
    @State private var isSending: Bool = false
    @State private var error: String?
    @State private var sentSuccessfully: Bool = false
    @State private var recipientCount: Int = 0

    private let maxRecipients = 20
    private let currentUserId = AuthenticationService.shared.userId

    private var otherMembers: [CircleMemberInfo] {
        members.filter { $0.userId != currentUserId }
    }

    var body: some View {
        NavigationStack {
            Group {
                if sentSuccessfully {
                    successView
                } else {
                    composeView
                }
            }
            .background {
                ZStack {
                    TributeColor.charcoal.ignoresSafeArea()
                    TributeColor.warmGlow.ignoresSafeArea()
                }
            }
            .navigationTitle(sentSuccessfully ? "" : "SOS Prayer")
            .navigationBarTitleDisplayMode(.inline)
            .toolbarColorScheme(.dark, for: .navigationBar)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(sentSuccessfully ? "Done" : "Cancel") { dismiss() }
                        .foregroundStyle(.secondary)
                }
            }
        }
    }

    @ViewBuilder
    private var composeView: some View {
        ScrollView {
            VStack(spacing: 24) {
                VStack(spacing: 12) {
                    ZStack {
                        Circle()
                            .fill(TributeColor.warmCoral.opacity(0.12))
                            .frame(width: 64, height: 64)
                        Image(systemName: "exclamationmark.bubble.fill")
                            .font(.system(size: 26))
                            .foregroundStyle(TributeColor.warmCoral)
                    }

                    Text("Request Urgent Prayer")
                        .font(.system(.headline, design: .serif))
                        .foregroundStyle(TributeColor.warmWhite)

                    Text("Select up to \(maxRecipients) people who will be notified to pray for you.")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                        .multilineTextAlignment(.center)
                }

                VStack(alignment: .leading, spacing: 10) {
                    HStack {
                        Text("Message")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(.secondary)
                        Spacer()
                        Text("\(message.count)/500")
                            .font(.caption2)
                            .foregroundStyle(.tertiary)
                    }

                    TextField("Please pray for me...", text: $message, axis: .vertical)
                        .textFieldStyle(.plain)
                        .lineLimit(3...6)
                        .padding(12)
                        .background(TributeColor.cardBackground)
                        .clipShape(.rect(cornerRadius: 10))
                        .overlay(
                            RoundedRectangle(cornerRadius: 10)
                                .strokeBorder(TributeColor.cardBorder, lineWidth: 0.5)
                        )
                        .onChange(of: message) { _, newValue in
                            if newValue.count > 500 {
                                message = String(newValue.prefix(500))
                            }
                        }
                }
                .padding(.horizontal, 20)

                VStack(alignment: .leading, spacing: 12) {
                    HStack {
                        Text("Recipients")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(.secondary)
                        Spacer()
                        Text("\(selectedRecipientIds.count)/\(maxRecipients)")
                            .font(.caption2.weight(.medium))
                            .foregroundStyle(selectedRecipientIds.count >= maxRecipients ? TributeColor.warmCoral : .secondary)
                    }

                    if otherMembers.isEmpty {
                        HStack(spacing: 8) {
                            Image(systemName: "person.slash")
                                .font(.caption)
                                .foregroundStyle(.secondary)
                            Text("No other members in this circle yet.")
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                        .padding(12)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(TributeColor.cardBackground)
                        .clipShape(.rect(cornerRadius: 10))
                    } else {
                        Button {
                            if selectedRecipientIds.count == min(otherMembers.count, maxRecipients) {
                                selectedRecipientIds.removeAll()
                            } else {
                                selectedRecipientIds = Set(otherMembers.prefix(maxRecipients).map(\.userId))
                            }
                        } label: {
                            let allSelected = selectedRecipientIds.count == min(otherMembers.count, maxRecipients)
                            HStack(spacing: 6) {
                                Image(systemName: allSelected ? "checkmark.circle.fill" : "circle")
                                    .font(.subheadline)
                                    .foregroundStyle(allSelected ? TributeColor.golden : .secondary)
                                Text(allSelected ? "Deselect All" : "Select All (\(min(otherMembers.count, maxRecipients)))")
                                    .font(.caption.weight(.medium))
                                    .foregroundStyle(.primary)
                            }
                        }

                        ForEach(otherMembers) { member in
                            recipientRow(member)
                        }
                    }
                }
                .padding(.horizontal, 20)

                if let error {
                    HStack(spacing: 6) {
                        Image(systemName: "exclamationmark.triangle.fill")
                            .font(.caption)
                        Text(error)
                            .font(.caption)
                    }
                    .foregroundStyle(TributeColor.warmCoral)
                    .padding(.horizontal, 20)
                }

                Button {
                    Task { await sendSOS() }
                } label: {
                    Group {
                        if isSending {
                            ProgressView()
                                .tint(TributeColor.charcoal)
                        } else {
                            HStack(spacing: 8) {
                                Image(systemName: "bolt.heart.fill")
                                Text("Send SOS Prayer Request")
                            }
                        }
                    }
                    .tributeButton(color: TributeColor.warmCoral)
                }
                .disabled(selectedRecipientIds.isEmpty || isSending)
                .opacity(selectedRecipientIds.isEmpty ? 0.5 : 1)
                .padding(.horizontal, 20)
            }
            .padding(.top, 16)
            .padding(.bottom, 40)
        }
    }

    @ViewBuilder
    private var successView: some View {
        VStack(spacing: 24) {
            Spacer()

            ZStack {
                Circle()
                    .fill(TributeColor.sage.opacity(0.12))
                    .frame(width: 88, height: 88)
                Image(systemName: "hands.sparkles.fill")
                    .font(.system(size: 36))
                    .foregroundStyle(TributeColor.sage)
            }

            VStack(spacing: 8) {
                Text("Prayer Request Sent")
                    .font(.system(.title3, design: .serif, weight: .bold))
                    .foregroundStyle(TributeColor.warmWhite)

                Text("\(recipientCount) \(recipientCount == 1 ? "person has" : "people have") been asked to pray for you.")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                    .multilineTextAlignment(.center)
            }

            Text("\"Bear one another's burdens, and so fulfill the law of Christ.\"")
                .font(.system(.subheadline, design: .serif))
                .italic()
                .foregroundStyle(TributeColor.softGold.opacity(0.6))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 32)

            Text("Galatians 6:2")
                .font(.caption)
                .foregroundStyle(TributeColor.golden.opacity(0.5))

            Spacer()
            Spacer()
        }
        .padding(.horizontal, 24)
    }

    private func recipientRow(_ member: CircleMemberInfo) -> some View {
        let isSelected = selectedRecipientIds.contains(member.userId)
        let isDisabled = !isSelected && selectedRecipientIds.count >= maxRecipients

        return Button {
            if isSelected {
                selectedRecipientIds.remove(member.userId)
            } else if selectedRecipientIds.count < maxRecipients {
                selectedRecipientIds.insert(member.userId)
            }
        } label: {
            HStack(spacing: 12) {
                ZStack {
                    Circle()
                        .fill(isSelected ? TributeColor.golden.opacity(0.15) : TributeColor.cardBackground)
                        .frame(width: 36, height: 36)
                    Image(systemName: "person.fill")
                        .font(.caption)
                        .foregroundStyle(isSelected ? TributeColor.golden : .secondary)
                }

                VStack(alignment: .leading, spacing: 2) {
                    Text("Member")
                        .font(.subheadline.weight(.medium))
                        .foregroundStyle(.primary)
                    if member.role == "admin" {
                        Text("Admin")
                            .font(.caption2)
                            .foregroundStyle(TributeColor.golden)
                    }
                }

                Spacer()

                Image(systemName: isSelected ? "checkmark.circle.fill" : "circle")
                    .font(.title3)
                    .foregroundStyle(isSelected ? TributeColor.golden : Color.white.opacity(0.15))
            }
            .padding(10)
            .background(isSelected ? TributeColor.golden.opacity(0.04) : TributeColor.cardBackground)
            .clipShape(.rect(cornerRadius: 10))
            .overlay(
                RoundedRectangle(cornerRadius: 10)
                    .strokeBorder(isSelected ? TributeColor.golden.opacity(0.2) : TributeColor.cardBorder, lineWidth: 0.5)
            )
        }
        .disabled(isDisabled)
        .opacity(isDisabled ? 0.4 : 1)
    }

    private func sendSOS() async {
        isSending = true
        error = nil

        let sosMessage = message.trimmingCharacters(in: .whitespaces)
        let finalMessage = sosMessage.isEmpty ? "Please pray for me" : sosMessage

        do {
            let response = try await APIService.shared.sendSOS(
                circleId: circleId,
                message: finalMessage,
                recipientIds: Array(selectedRecipientIds)
            )
            recipientCount = response.recipientCount
            withAnimation(.spring(duration: 0.5)) {
                sentSuccessfully = true
            }
        } catch {
            self.error = error.localizedDescription
        }
        isSending = false
    }
}
